# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- epilogue_with_visitor.h
+++ epilogue_with_visitor.h
@@ -42,6 +42,8 @@
 #include "cutlass/semaphore.h"
 #include "cutlass/epilogue/threadblock/epilogue_base.h"
 
+#include "sync.h"
+
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -275,12 +277,12 @@ public:
       // Convert and store fragment
       //
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       acc2smem_source_needed<cutlass::make_index_sequence<Visitor::kIterations>>::push(
           iter_idx, accum_fragment_iterator, this->warp_tile_iterator_);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
