# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- epilogue.h
+++ epilogue.h
@@ -62,6 +62,8 @@
 #include "cutlass/epilogue/threadblock/predicated_tile_iterator.h"
 #include "cutlass/numeric_types.h"
 
+#include "sync.h"
+
 ////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -273,7 +275,7 @@ private:
       // Convert and store fragment
       //
       
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
 
       acc2smem_source_not_needed<
@@ -282,7 +284,7 @@ private:
                                                                         accum_fragment_iterator,
                                                                         this->warp_tile_iterator_);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
@@ -399,13 +401,13 @@ private:
       //
       // Convert and store fragment
       //
-      
-      __syncthreads();
+
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       acc2smem_source_needed<cutlass::make_index_sequence<OutputTileIterator::kIterations>>::push(
           iter, accum_fragment_iterator, this->warp_tile_iterator_);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
