# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- epilogue_with_broadcast.h
+++ epilogue_with_broadcast.h
@@ -68,6 +68,8 @@
 
 #include "cutlass/numeric_types.h"
 
+#include "sync.h"
+
 /////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -536,7 +538,7 @@ private:
       //
       
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       acc2smem_source_not_needed<
           cutlass::make_index_sequence<OutputTileIterator::kIterations /
@@ -544,7 +546,7 @@ private:
                                                                         accum_fragment_iterator,
                                                                         this->warp_tile_iterator_);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
@@ -677,12 +679,12 @@ private:
       // Convert and store fragment
       //
       
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       acc2smem_source_needed<cutlass::make_index_sequence<OutputTileIterator::kIterations>>::push(
           iter, accum_fragment_iterator, this->warp_tile_iterator_);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
