# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- epilogue_planar_complex.h
+++ epilogue_planar_complex.h
@@ -56,6 +56,8 @@
 #include "cutlass/epilogue/threadblock/epilogue_base.h"
 #include "cutlass/epilogue/threadblock/predicated_tile_iterator.h"
 
+#include "sync.h"
+
 ////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -285,7 +287,7 @@ public:
       // Convert and store fragment
       //
       
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       typename AccumulatorFragmentIterator::Fragment accum_fragment_real;
       typename AccumulatorFragmentIterator::Fragment accum_fragment_imag;
@@ -299,7 +301,7 @@ public:
       this->warp_tile_iterator_.store(accum_fragment_real);
       this->warp_tile_iterator_.store_with_pointer_offset(accum_fragment_imag, SharedStorage::kImaginaryStride);
 
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
 
       //
       // Load fragments from shared memory
