# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- mma_sparse_multistage.h
+++ mma_sparse_multistage.h
@@ -44,6 +44,8 @@
 
 #include "cutlass/gemm/threadblock/mma_sparse_base.h"
 
+#include "sync.h"
+
 /////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -476,7 +478,7 @@ public:
 
     // DEPBAR+SYNC
     cutlass::arch::cp_async_wait<Base::kStages - 2>();
-    __syncthreads();
+    ark::sync_warps<Base::WarpCount::kCount * 32>();
 
     // Pair of fragments used to overlap shared memory loads and math
     // instructions
@@ -596,7 +598,7 @@ public:
 
           // Waits until kStages-2 stages have committed. 
           arch::cp_async_wait<Base::kStages - 2>();
-          __syncthreads();
+          ark::sync_warps<Base::WarpCount::kCount * 32>();
 
           // Move to the next stage
           iterator_A.add_tile_offset({0, 1});
