# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- mma_pipelined.h
+++ mma_pipelined.h
@@ -45,6 +45,8 @@
 #include "cutlass/gemm/gemm.h"
 #include "cutlass/gemm/threadblock/mma_base.h"
 
+#include "sync.h"
+
 /////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -217,7 +219,7 @@ public:
     ++this->smem_iterator_A_;
     ++this->smem_iterator_B_;
 
-    __syncthreads();
+    ark::sync_warps<Base::WarpCount::kCount * 32>();
 
     // Pair of fragments used to overlap shared memory loads and math instructions
     WarpFragmentA warp_frag_A[2];
@@ -267,7 +269,7 @@ public:
 
           this->smem_iterator_B_.store(transform_B(tb_frag_B));
 
-          __syncthreads();
+          ark::sync_warps<Base::WarpCount::kCount * 32>();
           
           ++this->smem_iterator_A_;
           ++this->smem_iterator_B_;
