# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

--- mma_multistage.h
+++ mma_multistage.h
@@ -44,6 +44,8 @@
 
 #include "cutlass/gemm/threadblock/mma_base.h"
 
+#include "sync.h"
+
 /////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace cutlass {
@@ -427,7 +429,7 @@ public:
 
     // Waits until stages up to the previous (kStages-2)th stage have committed.
     cutlass::arch::cp_async_wait<Base::kStages - 2>();
-    __syncthreads();
+    ark::sync_warps<Base::WarpCount::kCount * 32>();
 
     // Pair of fragments used to overlap shared memory loads and math
     // instructions
@@ -556,7 +558,7 @@ public:
 
           // Waits until stages up to the previous (kStages-2)th stage have committed.
           arch::cp_async_wait<Base::kStages - 2>();
-          __syncthreads();
+          ark::sync_warps<Base::WarpCount::kCount * 32>();
 
           // Move to the next stage
           iterator_A.add_tile_offset({0, 1});
@@ -615,7 +617,7 @@ public:
       // commit and drain all pending and predicated LDGSTS pnz from the GEMM mainloop
       cutlass::arch::cp_async_fence();
       cutlass::arch::cp_async_wait<0>();
-      __syncthreads();
+      ark::sync_warps<Base::WarpCount::kCount * 32>();
     }
 
   }
